diff --git a/configure.ac b/configure.ac
index 32eae7889..5abe02bc1 100644
--- a/configure.ac
+++ b/configure.ac
@@ -227,6 +227,7 @@ ARG_ENABL_SET([xauth-eap],      [enable XAuth backend using EAP methods to verif
 ARG_ENABL_SET([xauth-pam],      [enable XAuth backend using PAM to verify passwords.])
 ARG_ENABL_SET([xauth-noauth],   [enable XAuth pseudo-backend that does not actually verify or even request any credentials.])
 # kernel interfaces / sockets
+ARG_DISBL_SET([ipsec-offload],  [disable the ipsec offload interface.])
 ARG_DISBL_SET([kernel-netlink], [disable the netlink kernel interface.])
 ARG_ENABL_SET([kernel-pfkey],   [enable the PF_KEY kernel interface.])
 ARG_ENABL_SET([kernel-pfroute], [enable the PF_ROUTE kernel interface.])
@@ -374,6 +375,7 @@ if test -z "$CFLAGS"; then
 fi
 AC_SUBST(PLUGIN_CFLAGS)
 AC_PROG_CC
+AC_PROG_CXX
 AM_PROG_CC_C_O
 
 AC_LIB_PREFIX
@@ -1603,6 +1605,7 @@ ADD_PLUGIN([kernel-libipsec],      [c charon cmd])
 ADD_PLUGIN([kernel-wfp],           [c charon])
 ADD_PLUGIN([kernel-iph],           [c charon])
 ADD_PLUGIN([kernel-pfkey],         [c charon nm cmd])
+ADD_PLUGIN([ipsec-offload],        [c charon nm cmd])
 ADD_PLUGIN([kernel-pfroute],       [c charon nm cmd])
 ADD_PLUGIN([kernel-netlink],       [c charon nm cmd])
 ADD_PLUGIN([selinux],              [c charon nm cmd])
@@ -1779,6 +1782,7 @@ AM_CONDITIONAL(USE_UPDOWN, test x$updown = xtrue)
 AM_CONDITIONAL(USE_DHCP, test x$dhcp = xtrue)
 AM_CONDITIONAL(USE_LOAD_TESTER, test x$load_tester = xtrue)
 AM_CONDITIONAL(USE_HA, test x$ha = xtrue)
+AM_CONDITIONAL(USE_IPSEC_OFFLOAD, test x$ipsec_offload = xtrue)
 AM_CONDITIONAL(USE_KERNEL_NETLINK, test x$kernel_netlink = xtrue)
 AM_CONDITIONAL(USE_KERNEL_PFKEY, test x$kernel_pfkey = xtrue)
 AM_CONDITIONAL(USE_KERNEL_PFROUTE, test x$kernel_pfroute = xtrue)
@@ -2127,6 +2131,7 @@ AC_CONFIG_FILES([
 	src/libcharon/plugins/unity/Makefile
 	src/libcharon/plugins/uci/Makefile
 	src/libcharon/plugins/ha/Makefile
+	src/libcharon/plugins/ipsec_offload/Makefile
 	src/libcharon/plugins/kernel_netlink/Makefile
 	src/libcharon/plugins/kernel_pfkey/Makefile
 	src/libcharon/plugins/kernel_pfroute/Makefile
diff --git a/src/libcharon/Makefile.am b/src/libcharon/Makefile.am
index fd88237fe..8d0d8609b 100644
--- a/src/libcharon/Makefile.am
+++ b/src/libcharon/Makefile.am
@@ -573,6 +573,14 @@ if MONOLITHIC
 endif
 endif
 
+if USE_IPSEC_OFFLOAD
+  SUBDIRS += plugins/ipsec_offload
+if MONOLITHIC
+  libcharon_la_LIBADD += plugins/ipsec_offload/libstrongswan-ipsec-offload.la
+endif
+endif
+
+
 if USE_KERNEL_NETLINK
   SUBDIRS += plugins/kernel_netlink
 if MONOLITHIC
